### PR TITLE
FlowdockNotify - Plugin

### PR DESCRIPTION
**Pull request by cerw**
**Contribution Type:** new plugin
**Primary Area:** plugins
**Link to Intent to Implement:** Flowdock Push based on Slack Example

**Description of change:** Added new Class FlowdockNotify

**Description of solution:**
Here is example of config: 
```yml
flowdock_notify:
    api_key: "907b7073a30c......"
    email: "hello@nanosolutions.io"
```

Let me know what else I need to polish. 

Thanks

### CODE REVIEW

File: PHPCI/Plugin/FlowdockNotify.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            ->setSource("PHPCI")
+            ->setFromAddress($this->email)
+            ->setFromName($this->build->getProject()->getTitle())
+            ->setSubject($status)
+            ->setTags(['#ci'])
+            ->setLink($this->build->getBranchLink())
+            ->setContent($message);
+
+        if (!$push->sendTeamInboxMessage($flowMessage, array('connect_timeout' => 5000, 'timeout' => 5000))) {
+            // handle errors...
```
Comment (mavimo) [inline PHPCI/Plugin/FlowdockNotify.php]: Remove unrequired comment

File: PHPCI/Plugin/FlowdockNotify.php [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            throw new \Exception('Flowdock Failed :'.$flowMessage->getResponseErrors());
```
Comment (mavimo) [inline PHPCI/Plugin/FlowdockNotify.php]: Use sprintf to generate the exception message, eg:

    throw new \Exception(sprintf('Flowdock Failed: %s', $flowMessage->getResponseErrors()));

File: PHPCI/Plugin/FlowdockNotify.php [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+     * Run the Flowdock plugin.
+     * @return bool
+     * @throws \Exception
+     */
+    public function execute()
+    {
+
+        $message = $this->phpci->interpolate($this->message);
+
+        $successfulBuild = $this->build->isSuccessful();
```
Comment (mavimo) [inline PHPCI/Plugin/FlowdockNotify.php]: Use ternary operator

```php
$successfulBuild = $this->build->isSuccessful() ? 'Success' : 'Failed';
```

File: PHPCI/Plugin/FlowdockNotify.php [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+        if ($successfulBuild) {
```
Comment (mavimo) [inline PHPCI/Plugin/FlowdockNotify.php]: This isn't required since we are extracting the info above.

File: PHPCI/Plugin/FlowdockNotify.php [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+     * @param Build   $build
+     * @param array   $options
+     * @throws \Exception
+     */
+    public function __construct(Builder $phpci, Build $build, array $options = array())
+    {
+        $this->phpci = $phpci;
+        $this->build = $build;
+
+        if (is_array($options) && isset($options['api_key'])) {
```
Comment (mavimo) [inline PHPCI/Plugin/FlowdockNotify.php]: Change approach to make code more readable and remove 2 level nesting.
```
if (!is_array($options) || !isset($options['api_key'])) {
  throw new \Exception('Please define the api_key for flowdock_notify plugin!');
}

// .. other code
```

File: PHPCI/Plugin/FlowdockNotify.php [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+            if (isset($options['message'])) {
+                $this->message = $options['message'];
+            } else {
+                $this->message = 'Build %BUILD% has finished ';
+                $this->message .= 'for commit <a href="%COMMIT_URI%">%SHORT_COMMIT%</a> (%COMMIT_EMAIL%)> ';
+                $this->message .= 'on branch <a href="%BRANCH_URI%">%BRANCH%</a>';
+            }
+
+            if (isset($options['email'])) {
```
Comment (mavimo) [inline PHPCI/Plugin/FlowdockNotify.php]: Use ternary operator:
```php
$this->email = isset($options['email']) ? $options['email'] : 'PHPCI';
```

File: PHPCI/Plugin/FlowdockNotify.php [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        $this->phpci = $phpci;
+        $this->build = $build;
+
+        if (is_array($options) && isset($options['api_key'])) {
+            $this->api_key = trim($options['api_key']);
+
+            if (isset($options['message'])) {
+                $this->message = $options['message'];
+            } else {
+                $this->message = 'Build %BUILD% has finished ';
```
Comment (mavimo) [inline PHPCI/Plugin/FlowdockNotify.php]: Move default message in the class constant:
```php
class FlowdockNotify implements \PHPCI\Plugin {
    const MESSAGE_DEFAULT = 'Build %BUILD% has finished for commit <a href="%COMMIT_URI%">%SHORT_COMMIT%</a> (%COMMIT_EMAIL%)> on branch <a href="%BRANCH_URI%">%BRANCH%</a>';
    // ...
}
```

File: PHPCI/Plugin/FlowdockNotify.php [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    public function __construct(Builder $phpci, Build $build, array $options = array())
+    {
+        $this->phpci = $phpci;
+        $this->build = $build;
+
+        if (is_array($options) && isset($options['api_key'])) {
+            $this->api_key = trim($options['api_key']);
+
+            if (isset($options['message'])) {
+                $this->message = $options['message'];
```
Comment (mavimo) [inline PHPCI/Plugin/FlowdockNotify.php]: Use ternary operator:
```php
$this-> message = isset($options['message']) ? $options['message'] : self::MESSAGE_DEFAULT;
```

**Comment by cerw** [comment]
Hi
Thanks for feedback, always space to improve, I did the fixes you requested, let me know. 

Petr

**Comment by mavimo** [comment]
@cerw nice work! Should be possible to add some tests? I'm not sure should be easy since you instantiate the ```TeamInboxMessage``` into the execute method, but if possible should be good.

**Comment by cerw** [comment]
HI
I am gonna try do some test, but not many Plugins have many test from what i see..

**Comment by mavimo** [comment]
@cerw you are right write test for plugin should be complex (since the current code structure don't easily allow replacing part of code) but have tests is a really appreciate plus :)

**Pull request closed without merging**